### PR TITLE
Give a dedicated diagnostic for awaiting an `IAsyncEnumerable`

### LINKED ISSUE #53426: Improve error message for awaiting an IAsyncEnumerable value
The error message for misusing an API with IAsyncEnumerable isn't very helpful in figuring out the correct usage (need to use `await foreach`).
We can improve that easily at least in the case of a type implementing `IAsyncEnumerable`. We can at least address that scenario. 
The pattern-based scenario may be more difficult.

```
using System.Collections.Generic;
using System.Threading.Tasks;
public class C 
{
    public async Task M(IAsyncEnumerable<int> e) 
    {
        // error CS1061: 'IAsyncEnumerable<int>' does not contain a definition for 'GetAwaiter' and no accessible extension method 'GetAwaiter' accepting a first argument of type 'IAsyncEnumerable<int>' could be found (are you missing a using directive or an assembly reference?)
        await e; 
    }
}
```

### PR DESCRIPTION
**Pull request by nkarl7**
Fixes #53426

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=56885) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=56885) before we can accept your contribution.<br/><br/>:x: nkarl7 [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=56885)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=56885) it.</sub>

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 1]
```diff
@@ -2002,6 +2002,8 @@ internal enum ErrorCode
 
         #endregion
 
+        ERR_NoAwaitOnAsyncEnumerable = 8975
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:2005]: This error code is now used. You'll need to merge latest `main` and update it. Also, keep the trailing comma

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -6538,13 +6538,24 @@ private void BindMemberAccessReportError(BoundMethodGroup node, BindingDiagnosti
...
                     Error(diagnostics, ErrorCode.ERR_NoSuchMemberOrExtension, name, boundLeft.Type, plainName);
                 }
             }
         }
 
+        private bool ShouldBeUsingAwaitForeach(TypeSymbol receiver, string methodName)
+        {
+            // As per issue #53426: "Improve error message for awaiting an IAsyncEnumerable value".
+            return methodName == WellKnownMemberNames.GetAwaiter && receiver.IsIAsyncEnumerableType(Compilation);
+
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:6556]:
```suggestion
```

**Comment by Youssef1313** [comment]
Tagging @jcouv for visibility since this has been around for long without a review.

**Review by Youssef1313 — COMMENTED** [review]
> translations missing

You don't have to do anything here. The translations are done by the localization team.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 3]
```diff
@@ -8023,5 +8023,28 @@ public static class Extensions
...
+        await e; 
+    }
+}";
+
+            var expected = Diagnostic(ErrorCode.ERR_NoAwaitOnAsyncEnumerable, "await e")
+                .WithArguments("System.Collections.Generic.IAsyncEnumerable<int>", "GetAwaiter")
+                .WithLocation(8, 9);
+
+            CreateCompilationWithTasksExtensions(new[] { source, s_IAsyncEnumerable }, parseOptions: TestOptions.Regular8)
+                .VerifyDiagnostics(expected);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:8047]: We tend to use a different format for `VerifyDiagnostics`. See `TestWithCSharp7_3` for example. Something like:
```
            comp.VerifyDiagnostics(
                // (7,9): error CS8652: The feature 'async streams' is not available in C# 7.3. Please use language version 8.0 or greater.
                //         await foreach (int i in new C())
                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_3, "await").WithArguments("async streams", "8.0").WithLocation(7, 9)
                );
```

To achieve that, you can just let `VerifyDiagnostics` fail and copy&paste from the test failure output (it should contain the "expected" blurb).

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 4]
```diff
@@ -6538,13 +6538,24 @@ private void BindMemberAccessReportError(BoundMethodGroup node, BindingDiagnosti
...
                 {
                     Error(diagnostics, ErrorCode.ERR_NoSuchMemberOrExtension, name, boundLeft.Type, plainName);
                 }
             }
         }
 
+        private bool ShouldBeUsingAwaitForeach(TypeSymbol receiver, string methodName)
+        {
+            // As per issue #53426: "Improve error message for awaiting an IAsyncEnumerable value".
+            return methodName == WellKnownMemberNames.GetAwaiter && receiver.IsIAsyncEnumerableType(Compilation);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:6555]: >methodName == WellKnownMemberNames.GetAwaiter [](http://example.com/codeflow?start=19&length=45)

There's a test for method name. Please include a test that covers the negative outcome of that test (ie. cover all branches of the added logic).

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -6872,4 +6872,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="WRN_CompileTimeCheckedOverflow_Title" xml:space="preserve">
     <value>The operation may overflow at runtime (use 'unchecked' syntax to override)</value>
   </data>
+  <data name="ERR_NoAwaitOnAsyncEnumerable" xml:space="preserve">
+    <value>If you intend to iterate over an IAsyncEnumerable, then put the await statement before the loop statement. i.e. await foreach.</value>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:6876]: >If you intend to iterate over an IAsyncEnumerable, then put the await statement before the loop statement. i.e. await foreach. [](http://example.com/codeflow?start=11&length=126)

How about keeping the error portion of the message but change the suggestion portion?

`'{0}' does not contain a definition for '{1}' and no accessible extension method '{1}' accepting a first argument of type '{0}' could be found (did you mean to iterate over the async collection with `await foreach`?)`

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 6]
```diff
@@ -8023,5 +8023,28 @@ public static class Extensions
...
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:8048]: nit: consider adding a test for `IAsyncEnumerator`. The old error message should still be reported (can't `await foreach` on that type).

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1). Overall looks good. Some minor suggestions. Sorry about the delay in reviewing, somehow I missed it.

**Comment by jcouv** [comment]
@nkarl7 Sorry again for the delay reviewing this change. I'm going to mark the PR as draft for now, so that it doesn't show up in our overflowing review queue. Please re-publish/un-draft and ping when ready for another look. Thanks

**Pull request closed without merging**